sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RequestIdPolicy.java

```diff
@@ -3,22 +3,63 @@
 
 package com.azure.core.http.policy;
 
+import com.azure.core.http.HttpHeader;
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpRequest;
 import com.azure.core.http.HttpPipelineCallContext;
 import com.azure.core.http.HttpPipelineNextPolicy;
 import com.azure.core.http.HttpResponse;
 import reactor.core.publisher.Mono;
 
+import java.util.Objects;
 import java.util.UUID;
+import java.util.function.Supplier;
 
 /**
- * The pipeline policy that puts a UUID in the request header. Azure uses the request id as
- * the unique identifier for the request.
+ * The pipeline policy that adds request id header in {@link HttpRequest} once. These id does not change
+ * when request is retried. Azure uses the request id as the unique identifier for the request.
+ * Example of these headers are 'x-ms-client-request-id' and 'x-ms-correlation-request-id'.
  */
 public class RequestIdPolicy implements HttpPipelinePolicy {
     private static final String REQUEST_ID_HEADER = "x-ms-client-request-id";
+    private final Supplier<HttpHeaders> requestIdSupplier;
+
+    /**
+     * Creates default {@link RequestIdPolicy}.
+     */
+    public RequestIdPolicy() {
+        requestIdSupplier = () -> new HttpHeaders().put(REQUEST_ID_HEADER, UUID.randomUUID().toString());
+    }
+
+    /**
+     * Creates {@link RequestIdPolicy} with provided {@link Supplier} to dynamically generate request id for each
+     * {@link HttpRequest}.
+     *
+     * @param requestIdSupplier to dynamically generate to request id for each {@link HttpRequest}. It is suggested
+     * that this {@link Supplier} provides unique value every time it is called.
+     * Example of these headers are 'x-ms-client-request-id', 'x-ms-correlation-request-id'.
+     *
+     * @throws NullPointerException when {@code requestIdSupplier} is {@code null}.
+     */
+    public RequestIdPolicy(Supplier<HttpHeaders> requestIdSupplier) {
+        this.requestIdSupplier = Objects.requireNonNull(requestIdSupplier, "'requestIdSupplier' must not be null");
+    }
 
     @Override
     public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+
+        HttpHeaders httpHeaders = requestIdSupplier.get();
```

Comment:
Wouldn't this allow users to provide a random set of HTTP headers and not just the request id headers? 

For e.g. the supplier can return 
```java
new HttpHeaders().put("foo", "bar").put(REQUEST_ID_HEADER, UUID.randomUUID().toString());
```

Reply:
yes . But unless we keep a bag of valid request id header,  we can not validate. And whenever a new request-id header is introduced, we need to update the bag.